app/graphql/concerns/fancy_mutation.rb

```diff
@@ -0,0 +1,172 @@
+    # using the same error system as in #resolve. To achieve this, we prepend a module wrapping the
+    # mutation's #ready? and #authorized? methods, detecting the state of the error object after
+    # execution and changing the result if needed.
+    def ready?(*)
+      ready, result = super
+
+      return [false, { errors: errors }] if errors.present?
+
+      [ready, result]
+    end
+
+    def authorized?(*)
+      ready, result = super
+
+      return [false, { errors: errors }] if errors.present?
+
+      [ready, result]
+    end
+  end
+
+  class_methods do
+    # Sets the result type
+    # @param type [Class] The GraphQL type to use for the result
+    def result(type, **kwargs)
+      field :result, type, **kwargs
+    end
+
+    # Sets up an input type and field for the mutation
+    # @param block [Proc] A block defining the input type fields
+    def input(&block)
+      input_name = "#{graphql_name}Input"
+      input = Class.new(Types::Input::Base) do
+        graphql_name input_name
+        instance_eval(&block)
+      end
+      argument :input, input, required: true
+    end
+
+    # Adds types to the union of error types for the mutation
+    # @param types [Array<Class>] The error types this can return
+    def errors(*types)
+      errors_union.possible_types(*types)
+    end
+
+    # Adds types to the union of warning types for the mutation
+    # @param types [Array<Class>] The warning types this can return
+    def warnings(*types)
+      warnings_union.possible_types(*types)
+    end
+
+    # These two methods are kinda strange. The first time they're called, they create a new union
+    # type, hook it up to the mutation, and return it. When they get called again, they return that
+    # same type, allowing us to add more possible types to the union from different locations. This
+    # is important for utilities which might hook the `#ready?` method, since they can return
+    # errors.
+    def errors_union
+      @errors_union ||= begin
+        union_name = "#{graphql_name}ErrorsUnion"
+        errors_union = Class.new(Types::Union::Base) do
+          graphql_name union_name
+
+          def self.resolve_type(object, _context)
+            object.key?(:__type) ? object[:__type] : super
+          end
+        end
+        field :errors, [errors_union], null: true
+
+        errors_union
+      end
+    end
+
+    def warnings_union
+      @warnings_union ||= begin
+        union_name = "#{graphql_name}WarningsUnion"
+        warnings_union = Class.new(Types::Union::Base) do
+          graphql_name union_name
+
+          def self.resolve_type(object, _context)
+            object.key?(:__type) ? object[:__type] : super
+          end
+        end
+        field :warnings, [warnings_union], null: true
+
+        warnings_union
+      end
+    end
+  end
+
+  included do
+    prepend PrependedMethods
+    private_class_method :errors_union
+    private_class_method :warnings_union
```

Comment:
not seen this approach in a while.

Reply:
It's hella janky that I can't put this in class_methods but at least it still achieves what I want